mesonbuild/compilers/mixins/clike.py

```diff
@@ -925,13 +926,83 @@ def symbols_have_underscore_prefix(self, env: 'Environment') -> bool:
                     # Check if the underscore form of the symbol is somewhere
                     # in the output file.
                     if b'_' + symbol_name in line:
-                        mlog.debug("Symbols have underscore prefix: YES")
+                        mlog.debug("Underscore prefix check found prefixed function in binary")
                         return True
                     # Else, check if the non-underscored form is present
                     elif symbol_name in line:
-                        mlog.debug("Symbols have underscore prefix: NO")
+                        mlog.debug("Underscore prefix check found non-prefixed function in binary")
                         return False
-        raise RuntimeError(f'BUG: {n!r} check failed unexpectedly')
+        return None
+
+    def _symbols_have_underscore_prefix_define(self, env: 'Environment') -> T.Union[None,bool]:
+        '''
+        Check if symbols have underscore prefix by querying the
+        __USER_LABEL_PREFIX__ define that most compilers provide
+        for this. Return if funtions have underscore prefix or None
+        if it was not possible to determine, like when the compiler
+        does not set the define or the define has an unexpected value.
+        '''
+        delim = '"MESON_HAVE_UNDERSCORE_DELIMITER" '
+        code = f'''
+        #ifndef __USER_LABEL_PREFIX__
+        #define MESON_UNDERSCORE_PREFIX unsupported
+        #else
+        #define MESON_UNDERSCORE_PREFIX __USER_LABEL_PREFIX__
+        #endif
+        {delim}MESON_UNDERSCORE_PREFIX
+        '''
+        with self._build_wrapper(code, env, mode='preprocess', want_output=False, temp_dir=env.scratch_dir) as p:
+            if p.returncode != 0:
+                raise RuntimeError(f'BUG: Unable to preprocess _symbols_have_underscore_prefix_define check: {p.stdout}')
+            symbol_prefix = p.stdout.partition(delim)[-1].rstrip()
+
+            mlog.debug(f'Queried compiler for function prefix: __USER_LABEL_PREFIX__ is "{symbol_prefix!s}"')
+            if symbol_prefix == '_':
+                return True
+            elif symbol_prefix == '':
+                return False
+            else:
+                return None
+
+    def _symbols_have_underscore_prefix_list(self, env: 'Environment') -> T.Union[None,bool]:
+        '''
+        Check if symbols have underscore prefix by consulting a hardcoded
+        list of cases where we know the results.
+        Return if functions have underscore prefix or None if unknown.
+        '''
+        m = env.machines[self.for_machine]
+        # Darwin always uses the underscore prefix, not matter what
+        if m.is_darwin():
+            return True
+        # Windows uses the underscore prefix on x86 (32bit) only
+        if m.is_windows() or m.is_cygwin():
```

Comment:
I genuinely have no clue here, but...

Given cygwin is "generally like Linux instead of Windows", and it does not use the Windows libc -- the cygwin DLL is based on newlib -- is this still true?

Also curious what MSYS2 does here, although MSYS2 anyways patches meson for support (you cannot just pip install it).

Reply:
Given that this is about the __cdecl calling convention, I would be surprised if it would make a difference.  However I do not have any cygwin env to test it, I can try to set one up tomorrow though and check.

Reply:
We could write a test for it that involves using a third method of detecting symbols prefixes?

Generate a module with a `foo()` and `_foo()` and `__foo()` and have each of them print a different number when called. Write an executable that loads `foo` at runtime and see which one gets called when you run it?

This is a decent test but a bad configure test because it needs an exe wrapper when cross-compiling.

Reply:
The module would need to be generated with assembly or something that doesn't go through the linker's symbol mangling.

Reply:
Another option for the test is to run the `symbolextractor.py` script, which will give the mangled symbol list on each platform.

Reply:
Oh I see, Github confusingly split this thread up not showing the follow up comments before… 🤯 

Indeed for a unit test in meson we could just check the actual binary using a tool like `nm` or so to verify the list in a test. Or even simpler just rely on the current logic and make sure in the test we have no arguments that can mess with it. (Like -g and similar)

Reply:
Yeah, that's what I meant. We could write a unit test that uses methods from `symbolextractor.py` which already has a lot of platform-specific code. With that we won't need to worry about testing on cygwin by hand because the unit test will do that for us.

Reply:
This comment about the C ABI used in 32-bit Cygwin is correct.

I'd be surprised if we ever end up there anyhow, as any Cygwin-targetting compiler anyone might be using should be reporting __USER_LABEL_PREFIX__.

Reply:
I've added unittests now, please let me know if anything else is missing to get this merged.